### PR TITLE
Scenario builder for renewable [2/3]

### LINKED ISSUE #220: Feature wind off-shore (or miscellaneus renewable)
This is a new feature about a new kind of renewable (could be called misc. renewable).
More accurately, we want to offer the possibility to add renewable clusters (of wind or solar type) to any area and provide many time series (or TS) to each renewable cluster. For each renewable cluster and a given year, the related ready-made TS number is chosen by the user or drawn randomly.
 
## Antares main Window
- A new tab is to be added in the main notebook, named Renewable (or misc. renewable). 
- This tab is built on the model of the Thermal tab : it contains an area selector and a possibility to **add**/**delete** (possibly **clone**) renewable clusters to any area.
- Once an area is selected (as well as a cluster), 3 frames must appear : 
   - the list of renewable clusters (**up**),
   - a minimal frame containing some characteristics of the selected cluster : name, type [wind or solar], area it belongs to (**left**)
   - a table containing a time series per column (**right**)

### "Old" and "New" renewables

The existent **Wind** and **Solar** tabs are kept as they are now. To avoid confusion, a toggle button is to be added to select between "old" and "new" renewable.

Old renewables allow TS generation, but only "Solar" and "Wind".

New renewables do not allow TS generation, but multiple types of solar & wind.

## Scenario builder
The scenario builder (choice of TS number for a year and a cluster) is similar to the thermal scenario builder.

## Input
A new directory is created in the **input** directory, called **renewables**, with the same structure as the **thermal** directory, except for the **prepro** subdirectory, nonexistant here.
We have to read the content of this directory when we load the study, and write it when some cluster data changed or when we make a **save as** operation. 

### Building outputs
Regarding the renewable clusters, there are 2 types of results : 
- The classical **Wind** and **Solar** results. These results can be found only in the result files whose names are "values" prefixed. To the classical results (the TS chosen randomly or by user), we need to add the TS chosen for the renewable clusters of the same type : for example, for a given area, we add to the **Wind** classical results, the chosen TS associated to all renewable clusters of type **Wind** in the area.
So we only have to make an addition to variable **Wind** and **Solar** before they are printed. Should be short and easy.
- A new kind of results : the details of each renewable clusters, as it is done for the thermal clusters. Of course, these results are based on the TS chosen before the simulation, whether they were chosen by user (via scenario builder) or randomly. 

### Reading outputs
- Tab **General values**
Reading these file is already done. Nothong to do.
- Tab **Renewable values**
This should be a new tab, close to the **Thermal plants** tab and similar to it : it contains the details of every renewable cluster.

## Solver
- Data structure : we provide a data structure to contain the production of each renewable cluster.
**Caution** : we may have to refactor the already existing cluster class, so that it contains a common part (parent class),  a  thermal cluster part (derived from the parent class), and a renewable cluster class (also derived from the parent class).
- In case the scenario builder is not used for any renewable cluster, we have to choose randomly which TS is associated to that cluster.
-  As it is already done for **Wind** and **Solar**,  we must substract from the load, the production of the renewable clusters

### PR DESCRIPTION
**Pull request by flomnes**
**DONE**: user-specified scenarii

**TODO**: configure intra/inter-modal TS selection

### CODE REVIEW

File: src/libs/antares/study/parameters.cpp [hunk 1]
```diff
@@ -217,13 +219,15 @@ void Parameters::reset()
...
     nbTimeSeriesThermal = 1;
+    nbTimeSeriesRenewable = 1;
     // Time-series refresh
     timeSeriesToRefresh = 0; // None
     refreshIntervalLoad = 100;
     refreshIntervalSolar = 100;
     refreshIntervalHydro = 100;
     refreshIntervalWind = 100;
     refreshIntervalThermal = 100;
+    refreshIntervalRenewable = 100;
```
Comment (guilpier-code) [inline src/libs/antares/study/parameters.cpp]: **refreshIntervalRenewable** not needed : this parameter seems to be related only for stochastic TS, and renewable clusters don't have such time series

File: src/libs/antares/study/parameters.h [hunk 2]
```diff
@@ -280,6 +282,8 @@ class Parameters final
     uint refreshIntervalWind;
     //! Refresh interval (in years) for timeSeries : Thermal
     uint refreshIntervalThermal;
+    //! Refresh interval (in years) for timeSeries : Renewable
```
Comment (guilpier-code) [inline src/libs/antares/study/parameters.h]: Same remark : **refreshIntervalRenewable** regards only stochastic TS, and renewable clusters don't have such TSs

File: src/solver/simulation/timeseries-numbers.cpp [hunk 3]
```diff
@@ -187,6 +190,23 @@ static unsigned int CheckMatricesWidth(const Data::Study& study)
             }
             break;
         }
+        case Data::timeSeriesRenewable:
```
Comment (guilpier-code) [inline src/solver/simulation/timeseries-numbers.cpp:193]: Very strange function (I mean **CheckMatricesWidth**) : it runs a loop on areas (with a possible nested loop on clusters on thermal or renewable energy), gets the number of TSs for each, but seems to only use the last area / cluster.
From my point of view, something is absurd here.

File: src/solver/simulation/timeseries-numbers.cpp [hunk 4]
```diff
@@ -263,30 +286,34 @@ bool TimeSeriesNumbers::Generate(Data::Study& study)
...
+    uint32 draw[Data::timeSeriesCount] = {0, 0, 0, 0, 0, 0};
 
     const bool tsgen[Data::timeSeriesCount]
       = {0 != (Data::timeSeriesLoad & parameters.timeSeriesToRefresh),
          0 != (Data::timeSeriesHydro & parameters.timeSeriesToRefresh),
          0 != (Data::timeSeriesWind & parameters.timeSeriesToRefresh),
          0 != (Data::timeSeriesThermal & parameters.timeSeriesToRefresh),
-         0 != (Data::timeSeriesSolar & parameters.timeSeriesToRefresh)};
+         0 != (Data::timeSeriesSolar & parameters.timeSeriesToRefresh),
+         0 != (Data::timeSeriesRenewable & parameters.timeSeriesToRefresh)};
```
Comment (guilpier-code) [inline src/solver/simulation/timeseries-numbers.cpp]: ok, last element of **tsgen** will always be **false**, but I can't a way to do something else than that.

File: src/solver/simulation/timeseries-numbers.cpp [hunk 5]
```diff
@@ -316,20 +343,42 @@ bool TimeSeriesNumbers::Generate(Data::Study& study)
...
+                          = DRAW_A_RANDOM_NUMBER(Data::timeSeriesThermal,
+                                                 cluster->series->series,
+                                                 parameters.nbTimeSeriesThermal);
+                    }
                 }
-                else
+            }
+            // Renewables
+            {
+                auto end = area.renewable.list.mapping.end();
```
Comment (guilpier-code) [inline src/solver/simulation/timeseries-numbers.cpp]: **mapping** will disappear as a member of a list of renewable clusters

File: src/solver/simulation/timeseries-numbers.cpp [hunk 6]
```diff
@@ -316,20 +343,42 @@ bool TimeSeriesNumbers::Generate(Data::Study& study)
...
-                                             parameters.nbTimeSeriesThermal);
+                    auto* cluster = i->second;
+                    if (!cluster->enabled)
+                    {
+                        study.runtime->random[Data::seedTimeseriesNumbers].next();
+                    }
+                    else
+                    {
+                        cluster->series->timeseriesNumbers.entry[0][y]
+                          = DRAW_A_RANDOM_NUMBER(Data::timeSeriesRenewable,
```
Comment (guilpier-code) [inline src/solver/simulation/timeseries-numbers.cpp:378]: This should **not** change the reference results because only the TS's seed is changed, and after the draws for load to thermal.

File: src/libs/antares/study/parameters.h [hunk 7]
```diff
@@ -258,6 +258,8 @@ class Parameters final
     uint nbTimeSeriesWind;
     //! Nb of timeSeries : Thermal
     uint nbTimeSeriesThermal;
+    //! Nb of timeSeries : Renewable
+    uint nbTimeSeriesRenewable;
```
Comment (guilpier-code) [inline src/libs/antares/study/parameters.h]: We have a problem with **nbTimeSeriesRenewable** here : in the current state of the code, it will always be 1, as it is initialized to 1 and never set. Indeed, its thermal counterpart is the number of generated TS, and renewable clusters should not be concerned with this.
The problem is that this parameter is used elsewhere, for instance in the TS numbers draws.
A closer look should be given to this.
And a question is : if thermal has ready made TS, how does the TS number are drawn ?

File: src/solver/simulation/timeseries-numbers.cpp [hunk 8]
```diff
@@ -316,20 +343,42 @@ bool TimeSeriesNumbers::Generate(Data::Study& study)
...
+                                                 cluster->series->series,
+                                                 parameters.nbTimeSeriesRenewable);
+                    }
                 }
             }
         });
```
Comment (guilpier-code) [inline src/solver/simulation/timeseries-numbers.cpp:384]: A bit further in the code from here, why don't we have a call of **BUILD_LOG_ENTRY** for reneawbles ?

File: src/solver/simulation/timeseries-numbers.cpp [hunk 9]
```diff
@@ -405,6 +466,12 @@ bool TimeSeriesNumbers::Generate(Data::Study& study)
                     tsNumbers = &(area.wind.series->timeseriesNumbers);
                 else if (intermodal[TS_INDEX(Data::timeSeriesHydro)])
                     tsNumbers = &(area.hydro.series->timeseriesNumbers);
+                else if (intermodal[TS_INDEX(Data::timeSeriesThermal)]
```
Comment (guilpier-code) [inline src/solver/simulation/timeseries-numbers.cpp:470]: Why do we need to add a test on **intermodal** parameter for thermal cluster ? The code is supposed to work fine without it

**Review by guilpier-code — CHANGES_REQUESTED** [review]

File: src/solver/simulation/timeseries-numbers.cpp (hunk 6 above)
Comment (flomnes) [reply src/solver/simulation/timeseries-numbers.cpp:378]: I disagree. From the 2nd iteration, random generation structure `study.runtime->random[Data::seedTimeseriesNumbers]` may have a different state **if at least one renewable cluster is disabled**, from these lines :
```cpp
                auto end = area.renewable.list.mapping.end();
                for (auto i = area.renewable.list.mapping.begin(); i != end; ++i)
                {
                    auto* cluster = i->second;
                    if (!cluster->enabled)
                    {
                        study.runtime->random[Data::seedTimeseriesNumbers].next();
                    }
```

On the other hand, `study.runtime->random[Data::timeSeriesRenewable]` is unrelated and has no effect on existing random number draws (load, etc.). On that I agree.

File: src/solver/simulation/timeseries-numbers.cpp (hunk 9 above)
Comment (flomnes) [reply src/solver/simulation/timeseries-numbers.cpp:470]: Without lines 467-470, when only thermal & renewable intemodal correlation is enabled, `tsNumbers = nullptr` and the assertion at l. 476 fails.

File: src/solver/simulation/timeseries-numbers.cpp (hunk 8 above)
Comment (flomnes) [reply src/solver/simulation/timeseries-numbers.cpp:384]: Done

File: src/solver/simulation/timeseries-numbers.cpp (hunk 5 above)
Comment (flomnes) [reply src/solver/simulation/timeseries-numbers.cpp]: Fixed

File: src/libs/antares/study/parameters.cpp (hunk 1 above)
Comment (flomnes) [reply src/libs/antares/study/parameters.cpp]: Fixed

File: src/libs/antares/study/parameters.h (hunk 2 above)
Comment (flomnes) [reply src/libs/antares/study/parameters.h]: Fixed

File: src/solver/simulation/timeseries-numbers.cpp (hunk 3 above)
Comment (flomnes) [reply src/solver/simulation/timeseries-numbers.cpp:193]: I've run a few mental experiments on some width lists, here are the results.

Assume `TimeSeriesT = Data::timeSeriesLoad`. Very similar for clusters (thermal or renewable), the per-cluster lists are just concatenated into a big one, for which we can use the same reasoning as here.

### W=[1,1,1,1]
Four areas, each with a matrix of width 1. The function returns `w=1`.
### W=[2,2,2,2]
Four areas, each with a matrix of width 2. The function returns `w=2` after all iterations.
### W=[1,2,3,4]
Four areas, with matrices of width 1, 2, 3, 4 respectively. At the 3rd iteration, return 0.
### W=[4,3,2,1]
Four areas, with matrices of width 4, 3, 2, 1 respectively. At the 2nd iteration, return 0.

In conclusion, the function returns 0 whenever two width are different and different than 1. Otherwise, return the common value.

0 is an error value :
```cpp
nbTimeseries = CheckMatricesWidth<TimeSeriesT>(study);
if (!nbTimeseries)
{
    logs.error() << "Intra-modal correlation: Constraint violation: The number of time-series "
                    "must be identical for all areas";
    return false;
}
```

Comment (flomnes) [reply src/solver/simulation/timeseries-numbers.cpp:193]: More investigations. I wrote a small Python function to reproduce the behavior
```python3
def f(L):
    w = 1
    for p in L:
        if p != w:
            if p != 1 and w != 1:
                return 0
            if p > 1:
                w = p
    return w
```
For the testing part :
```
f([1, 1, 1, 1]) = 1
f([2, 2, 2, 2]) = 2
f([1, 2, 3, 4]) = 0
f([4, 3, 2, 1]) = 0
f([0, 0]) = 1
f([0, 1, 0]) = 1
f([0, 1, 2]) = 2
f([0, 2, 1]) = 2
f([2, 1, 0]) = 0 # PROBLEM : a permutation should not affect the result
```
There is a problem (see comment) on permutations.

Also, the function returns 1 when all matrices are empty (width=0), which is expected.

File: src/libs/antares/study/parameters.h (hunk 7 above)
Comment (flomnes) [reply src/libs/antares/study/parameters.h]: >  And a question is : if thermal has ready made TS, how does the TS number are drawn ?

Here is the code when expanding the `DRAW_A_RANDOM_NUMBER` macro.

```cpp
if (intramodal[TS_INDEX(T)]) {
  cluster->series->timeseriesNumbers.entry[0][y] = draw[TS_INDEX(T)];
 } else {
  if (tsgen[TS_INDEX(T)]) {
    cluster->series->timeseriesNumbers.entry[0][y] = (uint32)(floor((double)(study.runtime->random[Data::seedTimeseriesNumbers].next()) * parameters.nbTimeSeriesThermal));
  } else {
    cluster->series->timeseriesNumbers.entry[0][y] = (uint32)(floor((double)(study.runtime->random[Data::seedTimeseriesNumbers].next()) * cluster->series->series.width));
  }
 }
```
When `tsgen[TS_INDEX(T)]=false`, TS generation is disabled, the ready-made series width is used.

I agree that since no TS generation is possible for renewable clusters, we should **always** use `cluster->series->series.width`.

Comment (flomnes) [reply src/libs/antares/study/parameters.h]: Note that
```cpp
if (intramodal[z])
{
    draw[z] = (uint32)(floor(study.runtime->random[Data::seedTimeseriesNumbers].next()
                             * nbTimeseries[z]));
}
```
where `nbTimeseries[z]` is obtained with function `CheckMatricesWidth` : 
- N if all matrices are of width N >= 1
- 1 if all matrices are of width 0
- 0 in error cases (inconsistent width)

File: src/solver/simulation/timeseries-numbers.cpp (hunk 6 above)
Comment (flomnes) [reply src/solver/simulation/timeseries-numbers.cpp:378]: More on this : adding a renewable cluster changes the random generation process, even if this cluster is within an area that is not connected to the network.

But this behavior is the same with thermal clusters. And to me (and @sgatto) this is not a problem.

Comment (guilpier-code) [reply src/solver/simulation/timeseries-numbers.cpp:378]: ok, so we should expect a change in the results (compared to current reference) after the first year of simulation with the addition of renewable clusters

**Review by guilpier-code — APPROVED** [review]

**Comment by flomnes** [comment]
Local merge in PR #267
